### DISCUSSION TITLE
Google Summer of Code 2023

### DISCUSSION THREAD

**Discussion by armantekinalp** [GazzolaLab/PyElastica#225] [Ideas]
# Google Summer of Code

## About Elastica

Elastica is a free and open-source software project for the simulation of assemblies of slender, one-dimensional bodies using Cosserat rod theory. Slender structures are ubiquitous in both nature and engineered systems at all different sizes, from DNA strands to space tethers. Elastica has been designed to model such slender structures, with a particular focus on muscular structures and soft robotics. We are focused on providing useful simulation tools to the robotics and biomechanics communities to model, control, and visualize how these slender structures evolve and interact. Some examples of what Elastica has been for can be found [here](https://www.cosseratrods.org/cosserat_rods/case-studies/).

### PyElastica
PyElastica is the Python implementation of Elastica. PyElastica is designed to be modular, extensible, and easy to use. It allows the user to model a collection of slender rods that are each subject to different external (i.e. gravity, friction, etc…) and internal (i.e. muscle torque) forces. Rods can account for self contact and can be combined to create assemblies of rods capable of modeling increasingly complex systems.  We are actively developing it and continually adding new features and performance improvements.

## Application advice
Before applying for a Google Summer of Code project, make sure you review our [contribution guidelines](https://github.com/GazzolaLab/PyElastica/blob/master/CONTRIBUTING.md). Applicants should have prior experience with Python 3. A basic understanding of scientific computing, in particular computational mechanics/numerical simulations, is preferred, however, lack of such knowledge should not hamper your ability to come up with a competitive proposal. 

We have provided some projects ideas below. Before applying, make please reach out to us to discuss your idea. Make an issue on the [PyElastica GitHub repo](https://github.com/GazzolaLab/PyElastica) . Projects that have reached out and discussed with us will be prioritized. If you have general questions about PyElastica, you can also comment below.

**Deliverables:** Projects with limited and well-defined scopes will be favored. We would prefer a project that is finished, documented, tested, and deployed by the end of the summer even if it means all the possible features are not implemented. 

# Ideas List

## Theme: Improving environmental interactions

> ### Project 1: Extending contact module capabilities
>
>**Skills required:** Python
>**Difficulty:** Intermediate  
>**Project size:** 350 hours (depending on the extent of proposed features)
>**Description:** *This project is one we are especially interested in.* Currently, in PyElastica we can model the contact of rods with a frictional plane or rigid cylinders and themselves. We would like to extend our contact module for the rod and any arbitrary shape. This project is focused on the implementation of a contact module that can take any STL mesh, and create the geometry in PyElastica. Then in the case of rod and imported shape collision contact forces will be applied to the rod. Here we would like to see the extensibility and also scalability of the implemented method.
>**Mentors:**  Arman Tekinalp, Noel Naughton

## Theme: Improving backend functionality and optimizing kernels

>### Project 2: Adding plugin functionality to PyElastica
>
>**Skills required:** Python
>**Difficulty:** Intermediate  
>**Project size:** 175 hours (depending on the extent of proposed features)
>**Description:** This project is focused on adding plug-in functionality to PyElastica. Currently, we are building packages that use PyElastica as a dependency such [magnetopyelastica](https://github.com/armantekinalp/MagnetoPyElastica) our magnetic rod simulator. However, we would like to make these new features as a part of PyElastica and install them as a plugin for example `pip install pyelastica [magnetic]`. You will work on adding module functionality to PyElastica while planning extensibility in the future. 
This [python page](https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/) and [medium thread](https://mwax911.medium.com/building-a-plugin-architecture-with-python-7b4ab39ad4fc) might be useful. 
>**Mentors:**  Arman Tekinalp, Yashraj Bhosale

>### Project 3: Optimizing backend kernels in PyElastica
>
>**Skills required:** Python
>**Difficulty:** Advanced  
>**Project size:** 350 hours (depending on the extent of proposed features)
>**Description:** This project is focused on optimizing the backend kernels of PyElastica. Currently, we are using Numba njit compilation for accelerating our code, but we would like to improve the speed of the backend code without sacrificing the modularity and extensibility of the code. As a part of this project, you will work on optimizing some of the kernels in the backend. We would like you to use other accelerators like `Jax` or binding `C++` code with `pybind`. All optimized kernels have to be profiled and tested at the unit test level and simulation level. 
>**Mentors:**  Arman Tekinalp, Yashraj Bhosale

## Theme: Improving visualization abilities
Currently, we do not have great tools for visualizing simulations. We have identified three different aspects of visualization that we think would make a good GSoC project. We think each is large enough to be its own but it might be possible to combine projects together if one solution satisfies the requirements of both. 

>### Project 4: Integration with 3D rendering software
>
>**Skills required:** Python, JSON, Blender or equivalent  
>**Difficulty:** Intermediate  
>**Project size:** 350 hours
>**Description:** This project should focus on developing a pipeline to export simulation data from a finished Elastica simulation to an open-source 3D rendering software (such as Blender) to create high-resolution images and videos of simulations. It should provide both an automatic method based on reasonable default parameters as well as the ability to customize the visualization in Blender. Documentation and tutorials for users will also be required. 
>**Mentors:** Arman Tekinalp, Noel Naughton, Seung Hyun Kim 

>### Project 5: CAD integration
>
>**Skills required:** Python, CAD software
>**Difficulty:** Advanced  
>**Project size:** 350 hours
>**Description:** This project is focused on implementing automatic import and setup of simulation geometry from CAD files using a neutral file format (such as *.stl). It will develop an interface to read in a CAD file, convert the geometry to an Elastica simulation, allow necessary simulation parameters that can not be defined in the CAD file to be specified, and have substantial error checking to make sure the user does not forget necessary properties. The ability to export an Elastica simulation into the same neutral CAD format should also be implemented. 
>**Mentors:** Arman Tekinalp, Noel Naughton, Seung Hyun Kim

## Theme: Logging, saving, and restarting simulations

>### Project 6: Saving and restart framework for simulations
>
>**Skills required:** Python, database frameworks
>**Difficulty:** Easy to Intermdiate
>**Project size:** 175 hours
>**Description:** Develop a framework for saving simulations both as checkpoints which can be relaunched if the simulation crashes and at the end of the simulation. Enough information about the simulation should be included that the simulation can be loaded and restarted from the saved data and utilities to do so should also be written. A filetype/database standard for the data should be identified (we are flexible about what type and will work with you to identify it) in order to allow interoperability with postprocessing data pipelines. 
>**Mentors:** Arman Tekinalp, Noel Naughton, Seung Hyun Kim

>### Project 7: Debugger/logger of simulation instabilities
>
>**Skills required:** Python, data vizualization
>**Difficulty:** Easy
>**Project size:** 175 hours
>**Description:** Often, if a simulation becomes unstable and crashes, it can be difficult to identify what aspect of the simulation caused the crash. This project should develop a logging/debugging tool that can be enabled to log, vizualize, and (bonus) automatically identify the part of the simulation that caused the instability. We can work with you to identify the metrics to track, but you will need to devlop the utilities to log and visualize the data in a useful manner. 
>**Mentors:** Arman Tekinalp, Noel Naughton

[//]: # (Collection of URLs.)

[link-readme-gallary]: https://github.com/skim0119/PyElastica/blob/assets_logo/assets/alpha_gallery.gif
[link-project-website]: https://cosseratrods.org

**Comment by Rahul-JOON** [comment]
Hello @armantekinalp @nmnaughton  , 
I am Rahul Joon, an ECE undergrad sophomore. I am interested in Project 1  “Extending contact module capabilities” as mentioned above. I am very glad to be a part of the python and pyelastica community. 

**Firstly here’s where I come from:**

I have been studying and implementing computer science since High School and am proficient in Data Structures and Algorithms (qualified [ACM-ICPC](https://icpc.global/) regionals during high school).
I have on-hands project experiences from personal projects to internships at SDC, USICT; ranging from python development to web & app development([GitHub](https://github.com/Rahul-JOON)).

**My progress so far:**

I have set up the pyelastca-master codebase and have a decent overview of the same.
I have set up Povray and CAD along with other rendering dependencies.
Studied Examples extension and already solved issue #169 (pull request - [#232](https://github.com/GazzolaLab/PyElastica/pull/232))

**Next Step(s)**:

I will be diving into elastica module itself to study the rods and rigid bodies and finally understand how the contact module has been implemented with current functionalities.

I would highly appreciate it if the project mentors could discuss with me, in brief, the project and the approach of implementation necessary.

Thanks & Regards,
Rahul Joon

**Reply by armantekinalp** [reply]
Hello @Rahul-JOON 

I think first going into contact module and understanding implementation is a good starting point. I would also suggest going through the literature and find how other open source libraries that uses model contact dynamics between different objects and planes to have an understanding of state of the art implementations. Couple of the modules I can suggest are [MuJoCo](https://mujoco.org) and [Drake](https://drake.mit.edu) . You can also think if these modules can be added as dependency in PyElastica, but it will require having a solid and robust interface between them and PyElastica.

Thanks,

Arman

**Reply by Rahul-JOON** [reply]
Hello @armantekinalp 

Sorry for this later reply, Nevertheless I invested time in studying the implementation of the contact module and Now I understand that it is the elastica.interaction.AnisotropicFrictionalPlane that handles the frictional interaction between rods and planes by calculating all the necessary vectors.
To extend the interaction module to any arbitrary shapes, we can either write its separate library or incorporate MuJoCo as API calls. I think both the MuJoCo and Drake (that you recommended) are very resourceful and have very extended documentation.
I would prefer the second option.
Let me know about any complications with my approach and could you please tell me some specific tasks about the projects so that I can craft a work plan?

Thanks & Regards
Rahul

**Reply by armantekinalp** [reply]
Hi @Rahul-JOON,

In addition to AnisotropicFrictionPlasen module I would also suggest checking the [External Contact](https://github.com/GazzolaLab/PyElastica/blob/master/elastica/joint.py#L860) module.

We would prefer having a separate contact library because of future compatibility and also license issues. But if you have a plan to fix these issues for example, if MuJoCo changes some functions in future how can we prevent the code in PyElastica to break, then we can also consider having MuJoCo functions.

In your work plan, you can explain what type of contact algorithm you will use? Will this contact algorithm can take any STL type mesh? If you are using existing library how are you planning to couple with PyElastica? How are you planing to validate your implementations? What kind of example will you provide for the first time users? For example you can mention, you will import two different STL meshes one for a rigid object and one for a surface and simulate rod contact with them. 

I hope this helps,

Best,

Arman

**Reply by Rahul-JOON** [reply]
Hello @armantekinalp,

I spent the last week figuring out ways to solve the mentioned tasks. Here's where I got:

1.) Stl mesh to Pyelastica geometry
 - I couldn't find any helpful tech other than 'numpy-stl' package; ['numpy-stl'](https://pypi.org/project/numpy-stl/) package can help us extract data from an STL file, but we will need to write the code to covert that in PyElastica geometry ourselves.
 - This also creates another problem that still we can manually code to covert that in known structures only like rods, cylinders, spheres, and planes. I imagine we might be able to create more complex structures using these as fundamental units.

2.) Contact algorithms between different objects
 - Though Mujoco has extended documentation, they haven't mentioned any equations while calculating necessary vectors, rather have mentioned their workflow [here](https://mujoco.readthedocs.io/en/latest/computation.html).
 - If we get already known shapes through an stl mesh, we can use the old libraries and simulate contact forces as in the External Contact module.
 - For any arbitrary shape(s), its must to write a brand new class that will compute the fundamental forces; the algorithms for fundamental forces are as per [here](https://mattia-lab.com/wp-content/uploads/2018/06/Gazzola_RSOS_2018.pdf), we will rather need to apply it for any shape.

Kindly guide me on how to proceed further, I will also like to start writing Gsoc proposal now if that's okay with you.
Looking forward to working with you.

Thanks & Regards
Rahul

**Reply by armantekinalp** [reply]
Hi @Rahul-JOON ,

I think you can go ahead and start writing the proposal. Please mention all possible challenges as well as how you are planning to attack and solve them as you did above.

Best,

Arman

**Reply by Rahul-JOON** [reply]
Hello @armantekinalp,

Thanks for the 'Go ahead'. I have linked my first draft of the proposal [here](https://docs.google.com/document/d/1yStoDQuS5Ef5pTQHUURRy2JMhJ_LvMhehJUb6m0q-j8/edit?usp=sharing), please give it a read. Any modifications/rectifications are most welcome.

Thanks & Regards
Rahul

**Reply by Rahul-JOON** [reply]
Hello @armantekinalp @nmnaughton,

May you provide some feedback regarding the proposal linked above? I wish to align my approach in the right direction as much as possible before I submit the proposal.

Thanks & Regards
Rahul

**Reply by armantekinalp** [reply]
Hi @Rahul-JOON  I have read through the proposal, and it looks good. You can go ahead with the application/submission. If you need anything more from us, let us know here.

**Reply by Rahul-JOON** [reply]
Hi @armantekinalp,

Thanks for reviewing, I will be finalizing my proposal for submission now.

**Comment by HarshKhilawala** [comment]
Dear @armantekinalp @bhosale2 ,

I am a newcomer to PyElastica and I am eager to contribute to the community. I am interested in the GSoC 2023 idea - Project: Adding plugin functionality to PyElastica, but before I begin working on my project proposal, I would like to gain a better understanding of the PyElastica codebase and how it functions. As a student developer, I am highly motivated to learn and grow in the tech industry. I was wondering if you could provide me with any guidance on how to get started with contributing to PyElastica. Your help would be greatly appreciated.

**Reply by armantekinalp** [reply]
Dear @HarshKhilawala ,

I would suggest first going through the examples to understand how to use PyElastica. Then you can go over our [Contributions guideline](https://github.com/GazzolaLab/PyElastica/blob/master/CONTRIBUTING.md) and as well as [issues page](https://github.com/GazzolaLab/PyElastica/issues) in PyElastica. You can take one of these issues and start working while exploring the PyElastica in more detail.

Best,

Arman

**Comment by jallowmusa** [comment]
Hello @armantekinalp 

How are doing my name is Musa A Jallow a native of the Gambia I am a student in the university of the Gambia am yet to do Python in school but I did some Online Course I want to  in project 1 (Extending Contact Module Capabilities. I would need your help please

**Comment by jallowmusa** [comment]
Thanks you so much I will work on it

**Comment by AsadNizami** [comment]
Hi, this is Asad Nizami. I'll be participating in GSoC '23 and want to apply to PyElastica. I am particularly interested in - Project 3: Optimizing backend kernels in PyElastica and I am preparing a proposal for the same. Please let me know if any other project is of urgent requirement. Also, any heads up is much appreciated. Thanks.
@armantekinalp @bhosale2

**Reply by bhosale2** [reply]
Hi Asad,

I think you can go ahead and start writing the proposal. Project 2 (adding plugin functionality) is also something of interest to us.

Best,

Yashraj

**Reply by AsadNizami** [reply]
Sure, so would it be possible to merge 2 medium-sized projects to make a large one? Please let me know. I am looking for a large project.
Thanks.

**Reply by bhosale2** [reply]
Hi Asad,
@armantekinalp and I feel that Project 3 itself is a large project according to GSOC guidelines. As such, we would suggest you stick to Project 3 for the proposal.

Thanks,
Yashraj

**Reply by AsadNizami** [reply]
Hi @bhosale2 @armantekinalp,

I have made a [proposal](https://docs.google.com/document/d/1jupQRyIajojpR0VmzSPp9qDQhzAG6Gss75y_QUmjKnU/edit) for the third project. Do let me know if there's anything to change.
Thanks.

**Reply by bhosale2** [reply]
HI @AsadNizami, @armantekinalp, and I have read through the proposal, and it looks good to us. You can go ahead with the application/submission. If you need anything more from us, let us know here.

**Reply by AsadNizami** [reply]
Ok, thanks for reviewing.

**Comment by armantekinalp** [comment]
Hi I suggest everyone participating GSOC to submit their proposal, deadline is 4th of April.

**Comment by Rahul-JOON** [comment]
Hello @armantekinalp, @nmnaughton, @Ali-7800,

Thank you for considering me for the opportunity to collaborate with the PyElatica organization as a GSoC contributor. I will try to do my absolute best and work with utmost sincerity.

Kindly provide me with links to the mailing list, dev chat (if any), and if possible maybe schedule a video chat meeting for interaction and discussion regarding the project.

I look forward to all the learnings that lie ahead.

Thanks & Regards
Rahul
